Report better errors for values parsed from a PostCSS stylesheet

nex3 opened the pull request:
- [ ] bugfix
- [x] feature
- [ ] refactor
- [ ] tests
- [ ] documentation
- [ ] metadata

### Breaking Changes?

- [ ] yes
- [x] no

If yes, please describe the breakage.

### Please Describe Your Changes

Allow the user to provide a PostCSS node as context for `parse()`, and use that to generate better error messages in `Node.error()`.

Fixes #98

---

TODO:

- [ ] tests
- [x] documentation
- [x] updating type declaration

Please let me know what the best way is to test something like this.

nex3 commented:
@shellscape friendly ping!

shellscape commented on lib/SubInput.js:
```diff
@@ -0,0 +1,55 @@
+
+// A PostCSS Input that exposes a substring of a larger Input as though it were
+// the entire text to be parsed.
+module.exports = class SubInput {
```
Without seeing how this is used (via tests) I can't comment on whether or not this is a good name for this. 

Also, why does this not inherit from `Input`?

shellscape commented on lib/index.js:
```diff
@@ -8,14 +8,51 @@
-    const input = new Input(css, options);
+    let input;
+    if (options.context) {
+      assert(options.lineInContext);
```
using `assert` outside of tests is an antipattern imho. When/if that fails, it won't provide the user with any kind of a meaningful error message.

use `RangeError` and describe in detail what the user/caller is missing.

shellscape commented on lib/index.js:
```diff
@@ -32,6 +69,24 @@ module.exports = {
+    });
+  },
+
+  parseAtRuleParams(rule) {
```
why is this here?

shellscape replied:
never got a reply on this

shellscape commented on lib/index.js:
```diff
@@ -32,6 +69,24 @@ module.exports = {
     return parser.root;
   },
 
+  parseDeclValue(decl) {
```
why is this here?

shellscape replied:
never got a reply on this

shellscape commented on lib/index.js:
```diff
@@ -8,14 +8,51 @@
   parse(css, options) {
-    const input = new Input(css, options);
+    let input;
+    if (options.context) {
```
`options.context` isn't a documented option (neither are `options.lineInContext` nor `columnInContext`) so I can't comment on them. documentation isn't only useful to end users after merge, it's a key component of new feature PR reviews.

nex3 commented:
Friendly ping again!

shellscape commented:
Any word on the CI failures?

nex3 commented:
I'm not sure exactly how the tests work, but it looks like they're verifying the internal structures of the parsed nodes, which are intentionally changing. Is there a way I should regenerate the expectations?

nex3 commented:
Ping again—this is a major sticking point for Google's use of `postcss-values-parser`. I'd really like to avoid having to move off this package, but if contributions can't be accepted quickly we may be forced to do so.

shellscape commented:
Heya. I'm currently putting in 60 hour weeks and taking care of our daughter who's home now due to the pandemic. Sorry that I can't move faster for your contribs. If you need to move away from this package no one will blame you, but pressure isn't a motivational right now.

nex3 commented:
Fair enough—family comes first.

shellscape commented:
So taking another look at CI, your tests are failing because the snapshots don't match. If you're not familiar with snapshot testing, I'd highly recommend getting up to speed on that, as you'll run across that pretty often in the ecosystem.

Ava's Snapshot Docs https://github.com/avajs/ava/blob/master/docs/04-snapshot-testing.md

Also pervasive, are Jest's snapshots https://jestjs.io/docs/en/snapshot-testing upon which Ava's are loosely based.

shellscape commented:
Normally I'm loathe to post consecutive replies, but they're addressing different things, so here we go:

I'm still not keen on the overall approach, but I'm willing to think more about it - if you add some tests.

Snapshots are handy here, so I can quickly view the differences between the old output and the output with your changes. I'm also not sure why `parseDeclValue` and `parseAtRuleParams` were added as exports. Taking a look at them and the util function they're using, they look like they should be external helpers that utilize the package. But because there aren't any tests on them, I can't visualize quickly how they're intending on being used and if they should be first-class exports. I don't same that to shame, only that the lack of tests is hurting my ability to quickly make a call on the PR. Your comment https://github.com/shellscape/postcss-values-parser/pull/102#pullrequestreview-359979755 showed exactly how you should construct a test for that, and you've already managed to identify how to capture the error content. All you need do is create a snapshot from that.

And of course most importantly, tests to show this parsing a value in a file, and passing the context. The `test/fixtures` directory is handy for keeping files around for tests.

nex3 commented:
Thanks for the tips! I've fixed the existing test failures and added new tests for error-handling in the context of a larger PostCSS document.

I really thought I'd written a reply about `parseDeclValue` and `parseAtRuleParams`; sorry about that! :sweat_smile: They're utility functions for the cases where the CSS value you're parsing is either a declaration's value or a parameter for an at-rule. I think these first-class utility functions are warranted for of a combination of two reasons:

1. These are by far the most common sources of CSS values that are parsed by this package.

2. Properly configuring the `lineInContext` and `columnInContext` options for these two options is a little tricky, since PostCSS doesn't automatically provide the source line/column for the start of the values (only the beginning of the entire declaration or at-rule).

Utility functions makes it much more likely that users will correctly configure parsing for these values, which in turn will mean that their error messages will be clearer and more accurate.

shellscape commented:
This project is on my list for open source today. Are you still waiting to use this functionality or have you or your team moved on to use an alterantive?

nex3 commented:
This is definitely high priority for us!
